### PR TITLE
ManualStrategy: allow each try to set interval before next try

### PR DESCRIPTION
**Pull request by pda**
### Manual interval strategy

Sometimes you only know the desired interval after each try, e.g. a rate-limited API may include a `Retry-After` header. For these cases, the manual strategy can be used, with another strategy as its default:

```Go
// manually specify interval during each try, defaulting to 10 seconds
manual := roko.NewManual(roko.Constant(10 * time.Second))

roko.NewRetrier(
  roko.WithStrategy(manual.Register()),
  roko.WithMaxAttempts(10),
).Do(func(r *roko.Retrier) error {

  response := apiCall() // may be rate limited

  if err := response.HTTPError(); err != nil {
    if response.Status == HttpTooManyRequests {
      if wait, err := strconv.Atoi(response.Header("Retry-After")); err != nil {
        r.SetNextInterval(r, wait * time.Second) // respect the API
      }
    }
    return err
  }
  return nil
})
```

---

Alternatively, this could be added to roko's core `Retrier`, as a sibling of `r.Break()` which is essentially the `WithMaxAttempts` equivalent of manually setting the interval… I spiked that in https://github.com/buildkite/roko/compare/main...manually-set-next-interval but didn't love it. It's a bit simpler to use, though. Not sure.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -118,6 +118,33 @@ In this case, the amount of time the retrier will wait between attempts depends
...
+  roko.WithStrategy(manual.Register()),
+  roko.WithMaxAttempts(10),
+).Do(func(r *roko.Retrier) error {
+
+  response := apiCall() // may be rate limited
+
+  if err := response.HTTPError(); err != nil {
+    if response.Status == HttpTooManyRequests {
+      if wait, err := strconv.Atoi(response.Header("Retry-After")); err != nil {
+        r.SetNextInterval(r, wait * time.Second) // respect the API
```
Comment (triarius) [inline README.md:139]:
```suggestion
        manual.SetNextInterval(r, wait * time.Second) // respect the API
```

Comment (triarius) [reply README.md:139]: It seems a bit unwieldy to have to reference both the strategy and the retrier to do this. I also don't like that the strategy lives outside the scope of the retrier.

**Review by triarius — COMMENTED** [review]
I kinda like the other one better. It's certainly easier to use, as evidenced by my suggestion.
Can you give some more reasons on why you didn't love it?

**Comment by pda** [comment]
> I kinda like the other one better. It's certainly easier to use, as evidenced by my suggestion.
> Can you give some more reasons on why you didn't love it?

Given roko has a pluggable strategy for delays, it felt more elegant to implement it as one of those, rather than adding more methods to the core.
I like how this strategy demonstrates the composability of strategies; e.g. `manual` with a default of `constant`.

With the other implementation, the relationship between the configured strategy (e.g. `exponential`) and the manually set interval was a bit less clear.

> It seems a bit unwieldy to have to reference both the strategy and the retrier to do this. I also don't like that the strategy lives outside the scope of the retrier.

Yeah that sucks a bit. We could solve part of that by adding something like `Retrier.Strategy()` that could be accessed from inside the user function. But to do anything interesting with it you'd need to type-assert it to a specific strategy, which would end up pretty ugly I think. (But maybe I need to embrace Go's type assertion syntax 😅). Oh and that might be tricky because the _strategy_ itself is just a method on the struct, which would be harder to access.

I think the main reason I didn't love the previous implementation was a small thing: I didn't like using `*int` to indicate “optional int”, i.e. there's a difference between manually setting the next interval to 0 and not having set the next interval. And I didn't like having to manually unset it after the attempt.

But I since figured out how to use the current attempt count to make the user-specified interval only valid for the current attempt. (And again, maybe I just need to embrace pointer-to-int as optional int 😅)

Maybe I'll revise the other branch and open a pull request with usage doc etc for comparison.

**Comment by pda** [comment]
Actually I also prefer the `manually-set-next-interval` approach in #5 now.
I slightly tweaked it, improved the tests, and added a README section.

**Review by triarius — APPROVED** [review]
LGTM. I, too, am no fan of pointers as optionals. With the landing of generics, we could use something like https://github.com/moznion/go-optional or the more comprehensive https://github.com/samber/mo instead. But that would risk the wrath of many gophers. I suspect those libraries are not that ergonomic to use in a language like go though.

**Review by triarius — CHANGES_REQUESTED** [review]
I meant to approve #5, not this 🤦‍♂️

**Pull request closed without merging**